### PR TITLE
feat(next): Sentry client to respect metrics opt-out

### PR DESCRIPTION
**Pull request by david1alvarez**
## Because

- Currently metrics are still collected from the user via sentry, regardless of their "opt-out" status.

## This pull request

- Checks the user's account status via the cart service, which sets a tag at the top level of the sentry client instance
- If the metricsOptOut tag is set with the value `true`, then sentry does not send events from anywhere on the front end

## Issue that this pull request solves

Closes: [# FXA-10398](https://mozilla-hub.atlassian.net/browse/FXA-10398)

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).

### CODE REVIEW

File: libs/payments/cart/src/lib/cart.service.ts [hunk 1]
```diff
@@ -69,14 +72,18 @@ export class CartService {
...
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      [accountCustomer, _] = await Promise.all([
+        this.accountCustomerManager
+          .getAccountCustomerByUid(args.uid)
+          .catch((error) => {
+            if (!(error instanceof AccountCustomerNotFoundError)) {
+              throw error;
+            }
+          }),
+        this.updateSentryForAccount(args.uid),
```
[the remark this answers is not in the document]
Comment (david1alvarez) [reply libs/payments/cart/src/lib/cart.service.ts:85]: It would evaluate every one yes, but would set the flag according to the user's `metricsOptOutAt` db field. If the user hasn't opted out of metrics, then the `metricsOptOut` sentry flag gets set to `false`, and the metrics get logged as expected

File: apps/payments/next/app/[locale]/[offeringId]/checkout/[interval]/[cartId]/layout.tsx [hunk 2]
```diff
@@ -52,9 +53,10 @@ export default async function RootLayout({
   const cartDataPromise = getCartAction(params.cartId);
   const l10n = getApp().getL10n(locale);
   const [cms, cart] = await Promise.all([cmsDataPromise, cartDataPromise]);
+  console.log('layout cart', cart);
```
Comment (StaberindeZA) [inline apps/payments/next/app/[locale]/[offeringId]/checkout/[interval]/[cartId]/layout.tsx:56]: Testing related? :)

File: apps/payments/next/app/[locale]/[offeringId]/checkout/[interval]/[cartId]/layout.tsx [hunk 3]
```diff
@@ -52,9 +53,10 @@ export default async function RootLayout({
...
 
   return (
-    <>
+    <Template cart={cart}>
```
Comment (StaberindeZA) [inline apps/payments/next/app/[locale]/[offeringId]/checkout/[interval]/[cartId]/layout.tsx:59]: Doesn't Next.js automatically wrap the page with `template`, similar to how it automatically wraps pages with `layout`?

That said, does `template` make sense here? Looking at the Next.js docs, it doesn't seem like the recommended usage for `template` applies here. Happy to discuss further.

> While less common, you might choose to use a template over a layout if you want:
> - Features that rely on useEffect (e.g logging page views) and useState (e.g a per-page feedback form).
> - To change the default framework behavior. For example, Suspense Boundaries inside layouts only show the fallback the first time the Layout is loaded and not when switching pages. For templates, the fallback is shown on each navigation.

File: libs/payments/cart/src/lib/cart.manager.ts [hunk 4]
```diff
@@ -101,6 +101,7 @@ export class CartManager {
         ...cart,
         id: cart.id.toString('hex'),
         uid: cart.uid ? cart.uid.toString('hex') : undefined,
+        metricsOptedOut: null,
```
Comment (StaberindeZA) [inline libs/payments/cart/src/lib/cart.manager.ts:104]: I'd expect `metricsOptedOut` not to be part of `cart.manager` since its main purpose is to help manage the Cart database table.

Since `metricsOptedOut` isn't part of the DB table, and shouldn't be, I wouldn't expect any changes here.

**Review by StaberindeZA — COMMENTED** [review]
Just a few comments, happy to discuss further. This looks good though.

File: libs/payments/cart/src/lib/cart.manager.ts (hunk 4 above)
Comment (david1alvarez) [reply libs/payments/cart/src/lib/cart.manager.ts]: bumped to the service layer

File: apps/payments/next/app/[locale]/[offeringId]/checkout/[interval]/[cartId]/layout.tsx (hunk 3 above)
Comment (david1alvarez) [reply apps/payments/next/app/[locale]/[offeringId]/checkout/[interval]/[cartId]/layout.tsx:58]: Just wrapped it in a `useEffect` callback, which is probably more correct anyhow! It seems like the intended use-case of a template to me though -- the need here is to have a layer that watches the `cart` variable, and then makes an update call to Sentry depending on the cart state

File: libs/payments/cart/src/lib/cart.service.ts [hunk 5]
```diff
@@ -286,13 +293,18 @@ export class CartService {
   /**
    * Fetch a cart from the database by ID
    */
-  async getCart(cartId: string): Promise<WithUpcomingInvoiceCart> {
+  async getCart(
+    cartId: string
+  ): Promise<WithUpcomingInvoiceCart & { metricsOptedOut: boolean }> {
```
Comment (StaberindeZA) [inline libs/payments/cart/src/lib/cart.service.ts:298]: Could you update the type `WithUpcomingInvoiceCart` with `metricsOptedOut`?

File: libs/payments/cart/src/lib/cart.service.ts [hunk 6]
```diff
@@ -141,13 +145,13 @@ export class CartService {
...
 
-    return cart;
+    return { ...cart, metricsOptedOut };
   }
 
   /**
    * Create a new cart with the contents of an existing cart, in the initial state.
    */
-  async restartCart(cartId: string): Promise<ResultCart> {
+  async restartCart(cartId: string): Promise<WithMetricsFlagCart> {
```
Comment (StaberindeZA) [inline libs/payments/cart/src/lib/cart.service.ts:154]: Could we revert the changes to `restartCart` and `setupCart`? 

It's not necessary to include `metricsOptedOut` as part of these methods, since they don't get used in the location these methods are called.

Not for this PR, but I'm actually tempted to change the return type of these methods to just `cart.id`, since that's the only value required by consumers of this method.

File: libs/payments/cart/src/lib/cart.service.spec.ts [hunk 7]
```diff
@@ -683,4 +717,17 @@ describe('CartService', () => {
       });
     });
   });
+
+  // TODO: implement before merge
```
Comment (StaberindeZA) [inline libs/payments/cart/src/lib/cart.service.spec.ts:721]: Testing this would be good :D

File: libs/payments/cart/src/lib/cart.service.ts (hunk 6 above)
Comment (david1alvarez) [reply libs/payments/cart/src/lib/cart.service.ts:154]: sounds good

File: libs/payments/cart/src/lib/cart.service.ts (hunk 5 above)
Comment (david1alvarez) [reply libs/payments/cart/src/lib/cart.service.ts:298]: Sure - updating the name to `WithInvoiceAndMetricsCart` to reflect this.

File: tsconfig.base.json [hunk 8]
```diff
@@ -26,6 +26,7 @@
     "paths": {
       "@fxa/payments/capability": ["libs/payments/capability/src/index.ts"],
       "@fxa/payments/cart": ["libs/payments/cart/src/index.ts"],
+      "@fxa/payments/cart/*": ["libs/payments/cart/src/lib/*"],
```
Comment (StaberindeZA) [inline tsconfig.base.json:29]: Is this necessary? We've been trying to use barrel files for exports from `libs` as much as possible.

File: libs/payments/ui/src/lib/client/components/MetricsWrapper.tsx [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+}: {
+  cart?: WithContextCart;
+  children: React.ReactNode;
+}) {
+  useEffect(() => {
+    cart &&
+      Sentry.getGlobalScope().setTag('metricsOptedOut', cart.metricsOptedOut);
+  }, [cart]);
+
+  return <div>{children}</div>;
```
Comment (StaberindeZA) [inline libs/payments/ui/src/lib/client/components/MetricsWrapper.tsx:23]: Are these `<div>`'s necessary?

Comment (StaberindeZA) [reply libs/payments/ui/src/lib/client/components/MetricsWrapper.tsx:23]: Probably causing the strange screenshot

File: libs/payments/ui/src/lib/client/components/MetricsWrapper.tsx [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  cart,
+  children,
+}: {
+  cart?: WithContextCart;
+  children: React.ReactNode;
+}) {
+  useEffect(() => {
+    cart &&
+      Sentry.getGlobalScope().setTag('metricsOptedOut', cart.metricsOptedOut);
+  }, [cart]);
```
Comment (StaberindeZA) [inline libs/payments/ui/src/lib/client/components/MetricsWrapper.tsx:21]: [This doesn't need to hold up merging this PR] Could this be narrowed down to just `metricsOptedOut` instead of changes on the entire cart?

**Review by StaberindeZA — APPROVED** [review]
r+wc

Just a few small change requests, biggest one probably being the UI change shown in the screenshot below.

image

File: tsconfig.base.json (hunk 8 above)
Comment (david1alvarez) [reply tsconfig.base.json]: nope! initially done in favor of package sizes, but removing from this PR

File: libs/payments/ui/src/lib/client/components/MetricsWrapper.tsx (hunk 9 above)
Comment (david1alvarez) [reply libs/payments/ui/src/lib/client/components/MetricsWrapper.tsx]: it absolutely was :) fixed
Screenshot 2024-10-02 at 9 37 49 AM

File: libs/payments/ui/src/lib/client/components/MetricsWrapper.tsx (hunk 10 above)
Comment (david1alvarez) [reply libs/payments/ui/src/lib/client/components/MetricsWrapper.tsx]: `cart?.metricsOptedOut` is strictly better here :) updating
